proposals/0189-sbpf-stricter-elf-headers.md

```diff
@@ -0,0 +1,131 @@
+---
+simd: '0189'
+title: SBPF stricter ELF headers
+authors:
+  - Alexander Meißner
+category: Standard
+type: Core
+status: Idea
+created: 2024-10-21
+feature: TBD
+extends: SIMD-0178, SIMD-0179
+---
+
+## Summary
+
+Imposes more restrictions on what is expected of ELF headers.
+
+## Motivation
+
+After the removal of relocations in SIMD-0178 the ELF layout could be massively
+be simplified by constraining it to a strict subset of what ELF otherwise
+allows. Doing so not only reduces the complexity of validator implementations
+but also reduces the attack surface.
+
+## Alternatives Considered
+
+Moving away from ELF as a container format altogether. However this would only
+gain a very small file size advantage but otherwise loose all tooling
+compatibility.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+### File header
+
+The file size must not be less than `size_of::<Elf64Ehdr>()` (64 bytes),
+otherwise `ElfParserError::OutOfBounds` must be thrown.
+
+- `e_ident.ei_mag` must be `[0x7F, 0x45, 0x4C, 0x46]`
+- `e_ident.ei_class` must be `ELFCLASS64` (`0x02`)
+- `e_ident.ei_data` must be `ELFDATA2LSB` (`0x01`)
+- `e_ident.ei_version` must be `EV_CURRENT` (`0x01`)
+- `e_ident.ei_osabi` must be `ELFOSABI_NONE` (`0x00`)
+- `e_ident.ei_abiversion` must be `0x00`
+- `e_ident.ei_pad` must be `[0x00; 7]`
+- `e_type` must be `ET_DYN` (`0x0003`)
+- `e_machine` must be `EM_SBPF` (`0x0263`)
+- `e_version` must be `EV_CURRENT` (`0x00000001`)
+- `e_entry` is checked later (see dynamic symbol table)
+- `e_phoff` must be `size_of::<Elf64Ehdr>()` (64 bytes)
+- `e_shoff` is not checked
+- `e_flags` see SIMD-0161
+- `e_ehsize` must be `size_of::<Elf64Ehdr>()` (64 bytes)
+- `e_phnum` must not be less than `0x0005`
+- `e_phoff + e_phnum * size_of::<Elf64Phdr>()` must be less than the file size
+- `e_phentsize` must be `size_of::<Elf64Phdr>()` (56 bytes)
+- `e_shnum` is not checked
+- `e_shentsize` must be `size_of::<Elf64Shdr>()` (64 bytes)
+- `e_shstrndx` must be less than `e_shnum`
+
+If any check fails `ElfParserError::InvalidFileHeader` must be thrown.
+
+### Program headers
+
+|  purpose  |    p_type    |   p_flags  | p_vaddr |
+| --------- | ------------ | ---------- | ------- |
+| bytecode  | PT_LOAD      | PF_X       | 0 << 32 |
+| ro data   | PT_LOAD      | PF_R       | 1 << 32 |
+| stack     | PT_GNU_STACK | PF_R, PF_W | 2 << 32 |
+| heap      | PT_LOAD      | PF_R, PF_W | 3 << 32 |
```

Comment:
Why do we allow these headers? These regions are configured independently of the ELF. Do we allow these headers just so that we can reduce LLVM modifications? Just curious.

Reply:
It is so that the memory regions are correctly reflected in the ELF header. This makes it easier for tooling to detect issues (such as overlap in section allocation) and could simplify the validator implementation in the future.

Also, with the upcoming CPI syscall restrictions (necessary for direct mapping) we will limit pointers provided by the user to come from the stack and heap. Thus their layout will be further solidified.